subsys/bluetooth/audio/has.c

```diff
@@ -868,19 +856,102 @@ static int bt_has_cp_generic_update(struct has_client *client, uint8_t prev_inde
 	}
 }
 
+#if defined(CONFIG_BT_SETTINGS)
+struct client_context_store {
+	/* Last notified preset index */
+	uint8_t last_preset_index_known;
+};
+
+static int settings_set_cb(const char *name, size_t len_rd, settings_read_cb read_cb, void *cb_arg)
+{
+	struct client_context_store store;
+	struct client_context *context;
+	bt_addr_le_t addr;
+	const char *next;
+	ssize_t len;
+	int err;
+
+	LOG_DBG("%s", name);
+
+	if (!name) {
+		LOG_ERR("Insufficient number of arguments");
+		return -EINVAL;
+	}
+
+	err = bt_settings_decode_key(name, &addr);
+	if (err) {
+		LOG_ERR("Unable to decode address %s", name);
+		return -EINVAL;
+	}
+
+	settings_name_next(name, &next);
+
+	context = context_find(&addr);
+	if (context == NULL && len_rd) {
+		/* Find and initialize a free entry */
+		context = context_alloc(&addr);
+		if (context == NULL) {
+			LOG_ERR("Failed to allocate client_context for %s", bt_addr_le_str(&addr));
+			return -ENOMEM;
+		}
+
+		/* Notify all the characteristics values after reboot */
+		atomic_set(context->flags, BONDED_CLIENT_INIT_FLAGS);
+	}
+
+	if (len_rd) {
+		len = read_cb(cb_arg, &store, sizeof(store));
+		if (len < 0) {
+			LOG_ERR("Failed to decode value (err %zd)", len);
+			return len;
+		}
+
+		context->last_preset_index_known = store.last_preset_index_known;
+
+		LOG_DBG("Restored client_context for %s len %zd", bt_addr_le_str(&addr), len);
+	}
+
+	return 0;
+}
+
+BT_SETTINGS_DEFINE(has, "has", settings_set_cb, NULL);
+
+static void store_client_context(struct client_context *context)
+{
+	struct client_context_store store = {
+		.last_preset_index_known = context->last_preset_index_known,
+	};
+	int err;
+
+	LOG_DBG("%s last_preset_index_known 0x%02x",
+		bt_addr_le_str(&context->addr), store.last_preset_index_known);
+
+	err = bt_settings_store("has", 0, &context->addr, &store, sizeof(store));
+	if (err != 0) {
+		LOG_ERR("Failed to store err %d", err);
```

Comment:
If we fail to store this, should we retry?

Reply:
I see not retry attempts in the other places that call this function, thus I'd throw an error only